### DISCUSSION TITLE
How to access the dual variable of a convex optimization problem?

### DISCUSSION THREAD

**Discussion by 0xF0F0F0F** [yalmip/YALMIP#928] [Problems using YALMIP] [answered]
Hi,

I am trying to solve following problem and find the Lagrange multiplier for the 3rd inequality constraint:

y=sdpvar(N,1);
options = sdpsettings('solver','mosek');
% options = sdpsettings('solver','ipopt');
obj = y.'*H*y + f.'*y;
const=[A*y>=cl; A*y<=cu; gamma>=sqrt(y.'*R*y)];
optimize(const,obj,options);

both MOSEK and IPOPT could successfully find the optimal, and ipopt is slightly better. 

If I choose MOSEK, then I could get the Lagrange multiplier via "dual(const(3))" , however, if I choose IPOPT, "dual(const(3))" returns "NaN", How could I access the dual variable when using the ipopt?

Many Thanks

**Comment by johanlofberg** [comment, marked as answer]
Duals are not captured for general nonlinear solvers.

However, you should be careful with that model. The concave sqrt operator is implemented using an socp-representation, and this should not be possible in this case as you are using sqrt in a non convexity preserving fashion. It could be that you have R=I since YALMIP then automatically replaces that expression with norm(y) which then maps to an socp cone. However, that is not a good model for a general nonlinear solver, as YALMIP is forced to model the non-smooth socp cone for the nonlinear solver.

A better model for a nonlinear solver is most likely the convex and smooth model gamma^2 >= y'*R*y

If you want to use a square-root with a nonlinear solver, use sqrtm https://yalmip.github.io/squareroots

**Comment by johanlofberg** [comment]
https://yalmip.github.io/Extracting

Note though, solving this problem via ipopt via YALMIP is a bad idea normally except in trivial cases, as ipopt is relies heavily on receiving the hessian, and this is not done by YALMIP. fmincon typically performs better without hessians, but in this case where you can use mosek, that is what you should do.

**Reply by 0xF0F0F0F** [reply]
Thanks, the reason I need to use square root is to avoid scale problem, i.e. denominator of the Shape Ratio. I will change to the sqrtm, thanks again!

**Reply by johanlofberg** [reply]
You should use norm(R^0.5*y) (or any suitable low-rank factorization)

**Reply by johanlofberg** [reply]
i.e. use a conic solver, not a general nonlinear model.

If you really must use a nonlinear solver, I still think a quadratic form will perform better, possible with some trivial variable change such as e == R^0.5*y and then e'*e <= gamma^2

**Reply by 0xF0F0F0F** [reply]
I will try both, thank your for your replying! 💯
